Update/editor welcome tour tracking and fixes

autumnfjeld opened the pull request:
### Changes proposed in this Pull Request

![Screen Shot 2020-12-16 at 16 55 40](https://user-images.githubusercontent.com/5665959/102310683-98375a80-3fbf-11eb-810f-1364a4168aa3.png)


Add tracking to [Welcome Tour ](https://github.com/Automattic/wp-calypso/issues/47690)

- [x] When the tour modal opens
- [x] When a user dismisses the tour modal
- [x] When user minimizes the tour
- [x] For each slide, the slide number, whether it is the last slide and a unique string, such as the slide header
- [x] Tracking events for the thumbs up/down rating

### Testing instructions
* Sync Editing Toolkit to your sandbox (e.g. pull down this PR, then [sync to sandbox](PCYsg-ly5-p2#wpcom-sandbox)
* Add `define( 'SHOW_WELCOME_TOUR', true );`to your ` /home/wpcom/public_html/wp-content/mu-plugins/0-sandbox.php` file. This will force the Welcome Tour to show every time the editor loads.

* View the Welcome Tour model opening in the Editor 
* In DevTools Network request tab, verify tracking network request (see screenshot) for each of the tracking events.  And 
  * `calypso_editor_wpcom_tour_open`   
      - data: `{is_gutenboarding:}`
  * `calypso_editor_wpcom_tour_dismiss` 
      - data: `{is_gutenboarding: , slide_number: , action: }`
  * `calypso_editor_wpcom_tour_slide_view` (fired on each slide, but NOT after a maximize event)
     - data:  `{slide_number:, is_last_slide:, slide_heading:, is_gutenboarding:}`
  * `calypso_editor_wpcom_tour_minimize`
     - data `{slide_number:,  is_gutenboarding:}`
  * `calypso_editor_wpcom_tour_minimize`
     - data `{slide_number:,  is_gutenboarding:}`
  * `calypso_editor_wpcom_tour_rate`
       - data `{thumps_up: , is_gutenboarding:}`


Ensure the following during testing
* slide view after maximizing _should not_ be tracked
* slide view going backward _should be_ tracked

autumnfjeld committed:
WelcomeTourTracking: track open & dismiss, fix usage of portalParent by making it a state variable, implement early return correctly

autumnfjeld committed:
Track the slide number at which user dismisses the tour

autumnfjeld committed:
Welcome Tour tracking: track each slide view , both going forward and going back

ramonjd commented:
I know you're still working on this, but I took it for a spin. Looking good so far. The events match the current version.

### NUX modal opens

```
is_gutenboarding: false
_en: calypso_editor_wpcom_tour_open
```


### Slide navigation

For each slide, e.g.,:
```
slide_number: 2
is_last_slide: false
slide_heading: Everything is a block
is_gutenboarding: false
_en: calypso_editor_wpcom_tour_slide_view
```

### Last slide 
```
slide_number: 8
is_last_slide: true
slide_heading: Congratulations!
is_gutenboarding: false
_en: calypso_editor_wpcom_tour_slide_view
```

### User clicks on Done button

```
is_gutenboarding: false
slide_number: 8
_en: calypso_editor_wpcom_tour_dismiss
```

### User dismisses by clicking on X
```
is_gutenboarding: false
slide_number: 1
_en: calypso_editor_wpcom_tour_dismiss
```

### User dismisses by clicking on No thanks
```
is_gutenboarding: false
slide_number: 1
_en: calypso_editor_wpcom_tour_dismiss
```

### User clicks on Restart
```
slide_number: 1
is_last_slide: false
slide_heading: Welcome to WordPress!
is_gutenboarding: false
_en: calypso_editor_wpcom_tour_slide_view
```

### User minimizes the NUX modal
Nothing

### User maximizes the NUX modal
```
slide_number: 1
is_last_slide: false
slide_heading: Welcome to WordPress!
is_gutenboarding: false
_en: calypso_editor_wpcom_tour_slide_view
```

autumnfjeld committed:
Prevent tracking of a slide view if user has returned to slide by maximizing (returning from minimized state)

autumnfjeld committed:
Track when user minimizes the tour on a particular slide

autumnfjeld committed:
Seperate out card footer components in prep for last slide thumbs up/down component

autumnfjeld committed:
Update variable names to clarify cardIndex (array starting with 0) vs slide numbering (starting with 1)

autumnfjeld committed:
Add thumbsUp & thumbsDown svg icon

autumnfjeld committed:
Fix thumbs naming

autumnfjeld committed:
WelcomeTour last slide: add thumbs up/down UI and tracking

autumnfjeld commented:
@ramonjd  @dubielzyk  Do we want to count this (maximizing) as a "view card" event?  My thought is no, because the user just viewed the card before they minimized the tour. To me it would skew the tracking data.   I've set it up so a "view" of a card[ isn't tracked if the user has just remaximized](https://github.com/Automattic/wp-calypso/blob/d48cdfa5bff3717819785d2a9bfc5260d0d9ac18/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-welcome-tour%2Fsrc%2Ftour-card.js#L40) the Tour.    Let me know if we should handle this differently than I have done.

>  **User maximizes the NUX modal**
> ``` slide_number: 1
> is_last_slide: false
> slide_heading: Welcome to WordPress!
> is_gutenboarding: false
> _en: calypso_editor_wpcom_tour_slide_view ```

ramonjd commented:
> Do we want to count this (maximizing) as a "view card" event? My thought is no, because the user just viewed the card before they minimized the tour. 

Good question. I think it wouldn't hurt to have separate events for the min/max actions. That way, if it is at all interesting to anyone, we could study flows such as "user minimized the tour, did something, then maximized".

I had a similar question in relation to whether we should distinguish between "Done", "Close" and "No thanks" dismissal events. Technically, we could look at whether they viewed any slides at all before dismissing. Or perhaps we could add another property, e.g., `action/type/label` or something to the event to categorize them. For example:

### User clicks on Done button
```
is_gutenboarding: false
slide_number: 8
_en: calypso_editor_wpcom_tour_dismiss
action: 'done'
```

What do you think?

dubielzyk commented:
I'd also prefer to have separate events, but if it muddies the reporting I'd be hesitant. I'll leave it to you to decide whether it's safe to do so.

autumnfjeld committed:
Tour rating: Add disabled & active states to buttons ToDo: make state persistent to prevent multiple ratings

autumnfjeld committed:
WelcomeTour nux store: refactor and add tourRating to data store

autumnfjeld committed:
TourRating: update to store the user's rating and ensure they don't rate twice

autumnfjeld committed:
Tracking: on dismiss track the source of the dismiss action, either close-btn or no-thanks btn

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:
```diff
@@ -93,12 +149,59 @@ function CardOverlayControls( { onMinimize, onDismiss } ) {
+		<>
+			<p className="welcome-tour__end-text">Did you find this guide helpful?</p>
+			<div>
+				<Button
```
We should add an `aria-label` to these buttons. The icons are set as `aria-hidden` by Gutenberg, so currently these currently appear as label-less buttons in a screen reader.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:
```diff
@@ -93,12 +149,59 @@ function CardOverlayControls( { onMinimize, onDismiss } ) {
+			return;
+		}
+		isDisabled = true;
+		setTourRating( 'thumbs-up' );
```
Doesn't this means it's impossible to set `'thumbs-down'` in the data store.

autumnfjeld replied:
@p-jackson 
 > why do we remember the rating in the data store? Yes it allows us to remind the user of the selection they made last time they were in the tour popup, but only for this session.

You are correct to point this out. :) A few thoughts:

* "only for this session":  As far as I understand it, the value isn't just for this session.  It is set to [persist](https://github.com/Automattic/wp-calypso/blob/aedabce06259cf42585e3779eb25a59434780975/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-welcome-tour%2Fsrc%2Fstore.js#L59) and I observed the store value persisted after page reload.  Source: [Persistence Plugin](https://github.com/WordPress/gutenberg/blob/a82c257641739fcdec7ed910ad5ddf0cafee9658/packages%2Fdata%2Fsrc%2Fplugins%2Fpersistence%2FREADME.md#L4)
* I debated the best approach for this.  Yes, one implementation could have been to save state in the root component. A very reasonable approach, since right now the goal is just to prevent user from rating multiple times.
* In the future we will (assuming successful A/B test) also show the Tour from the More Menu (same as the current NUX ) and then we might want to know about previous rating, or whether to show the rating at all, or maybe we'd have a different rating question.  

I'd agree that I probably over designed it, and maybe should have simply set up state in a parent component, like `WelcomeTourFrame`, but the above was my thought process.

autumnfjeld replied:
> Doesn't this means it's impossible to set `'thumbs-down'` in the data store.

Ah! That's embarrassing!!!  I think I just hard coded that string for initial dev testing.  Fixed. Thanks for catching!!!

p-jackson replied:
> "only for this session": As far as I understand it, the value isn't just for this session. It is set to persist and I observed the store value persisted after page reload.

I missed this 🙇 Since this is persisted in local storage then your implementation looks the best.

I think I was imagining a design in my head where once you've rated the buttons might disappear and get replaced with a "thanks for rating" sort of message, and that felt more of a transient thing. But that's a design change, what you've got here looks good to me now that I understand the persistence thing.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-launch.js:
```diff
@@ -78,24 +89,31 @@ function WelcomeTourFrame() {
 
-function WelcomeTourMinimized( { onMaximize } ) {
+function WelcomeTourMinimized( { onMaximize, setJustMaximized } ) {
+	const handleOnMaximize = () => {
```
Looks like you also use the `onAction`/`handleAction` naming convention! I like it too :D Keep it up and we will eventually get it through the whole codebase 😅

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:
```diff
@@ -7,84 +7,140 @@ import './public-path';
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+
+// eslint-disable-next-line react-hooks/exhaustive-deps
+const useEffectOnlyOnce = ( func ) => useEffect( func, [] );
```
This is annoying, but I just tried it myself and my eslint set up is also complaining about using `[]` in the dependency list. I thought eslint wasn't supposed to complain when the array was empty because this is a well known pattern 😡

p-jackson replied:
No action item here, just me venting 😓

autumnfjeld replied:
My thoughts exactly.   It would be cool if there was a `useEffectOnlyOnce` in core, as I would think this is a sometimes necessary!

p-jackson replied:
Maybe `useOnMount()`? Since we are looking for that old `onMount()` behaviour in this case.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-launch.js:
```diff
@@ -18,8 +18,10 @@ import { Icon } from '@wordpress/icons';
+import { recordTracksEvent } from '@automattic/calypso-analytics';
 
 function LaunchWpcomWelcomeTour() {
+	const [ portalParent ] = useState( () => document.createElement( 'div' ) );
```
DOM elements aren't usually placed in "state". But I see that you don't want a new one created each render. Do you want `useRef` maybe instead of `useState`?

autumnfjeld replied:
Ah, thank you!!  I just went to do read up on my understanding of that and found [this](https://www.codebeast.dev/usestate-vs-useref-re-render-or-not/) super helpful:

1. useState causes re-render; useRef does not.
2. Both useState and useRef remembers their data after a re-render

p-jackson reviewed (commented):
Looking good. I've left some comments, but I also have the general comment of: why do we remember the rating in the data store? Yes it allows us to remind the user of the selection they made last time they were in the tour popup, but only for this session. After a page refresh we've forgotten. It sort of feels like a transient thing that we could keep in component state and forget about it after they're done.
There might be a future design coming up that's going to make use of it though, in which case ignore all that :)

autumnfjeld committed:
WelcomeTour rating: Fix assignment of thumbs-up and thumbs-down to properly record user selection

autumnfjeld committed:
WelcomeTour tracking: track maximize event calypso_editor_wpcom_tour_maximize

autumnfjeld committed:
Tracking: on dismiss track the source of the dismiss action, close-btn

autumnfjeld committed:
Remove button variation on last slide since it is replaced with the tour rating 👍 UI

autumnfjeld committed:
Change tracking prop from source to action

autumnfjeld committed:
Add aria labels

autumnfjeld committed:
WelcomeTour: Store parentPortal via useRef instead of useState

autumnfjeld commented:
@ramonjd @dubielzyk 



I've implemented both [minimizing](https://github.com/Automattic/wp-calypso/blob/b934442d8a8323534bfef12f11d77dfc0283d3b2/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-welcome-tour%2Fsrc%2Ftour-card.js#L129) and [maximizing](https://github.com/Automattic/wp-calypso/blob/c9ef856d0e029d79523e93b316c6e707fce8afe4/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-welcome-tour%2Fsrc%2Ftour-launch.js#L117) events .

I also added a [data prop to the dismiss events](https://github.com/Automattic/wp-calypso/blob/c9ef856d0e029d79523e93b316c6e707fce8afe4/apps%2Fediting-toolkit%2Fediting-toolkit-plugin%2Fwpcom-block-editor-welcome-tour%2Fsrc%2Ftour-launch.js#L79) to show where the user dismissed: currently we have two dissmiss possibilities: the [x] at the top right and the "No thanks".  The "Done" on the last slide no longer exists because the 👍🏼  rating lives in that area of the last card.

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-card.js:
```diff
@@ -7,98 +7,201 @@ import './public-path';
+					<CardNavigation
+						cardIndex={ cardIndex }
+						lastCardIndex={ lastCardIndex }
+						onDismiss={ () => onDismiss( 'close-btn' ) }
```
When clicking on "No thanks", we're sending `action: close-btn`.

It's because we're passing down this anonymous function to the child components. The anonymous function always calls `onDismiss( 'close-btn' )` and doesn't accept any arguments, e.g., `'no-thanks-btn'`

We should pass the function down instead:

```suggestion
						onDismiss={ onDismiss }
```

autumnfjeld replied:
Yes, don't know what I was thinking there.  Thanks for catching!

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/store.js:
```diff
@@ -2,10 +2,31 @@
 
-const reducer = ( state = {}, { type, isNuxEnabled } ) =>
-	'WPCOM_BLOCK_EDITOR_NUX_SET_STATUS' === type ? { ...state, isNuxEnabled } : state;
+const isNuxEnabledReducer = ( state = null, action ) => {
```
I have a feeling the default value of the state here should be `undefined`.

We're doing a type check over in [tour-launch.js](https://github.com/Automattic/wp-calypso/blob/fd46f55c4296c4c859c9cff582a2bc5692f0d16e/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-launch.js#L35) that explicitly checks for `undefined`. 

But if the default state value is `null`, it will never pass the test.

I noticed this when testing, and clearing my local storage/IndexedDB while sandboxed to force the tour to reopen.

autumnfjeld replied:
Yes....you are right. 👍

ramonjd reviewed (changes requested):
Looking great! Just one change request: we need to pass down `onDismiss` rather than the anonymous function so that the close action argument works.

Test steps:

**On a new site, the welcome tour opens and the following events fire:**

```
is_gutenboarding: true
_en: calypso_editor_wpcom_tour_open
```

```
slide_number: 1
is_last_slide: false
slide_heading: Welcome to WordPress!
is_gutenboarding: true
_en: calypso_editor_wpcom_tour_slide_view
```


**Clicking through slides**
```
slide_number: 2
is_last_slide: false
slide_heading: Everything is a block
is_gutenboarding: true
_en: calypso_editor_wpcom_tour_slide_view
```
...

```
slide_number: 8
is_last_slide: true
slide_heading: Congratulations!
is_gutenboarding: true
_en: calypso_editor_wpcom_tour_slide_view
```

**Minimized tour**

```
is_gutenboarding: true
slide_number: 1
_en: calypso_editor_wpcom_tour_minimize
```

**Maximized tour** 

```
is_gutenboarding: true
slide_number: 1
_en: calypso_editor_wpcom_tour_maximize
```

**Rating tour**
```
thumbs_up: true
is_gutenboarding: true
_en: calypso_editor_wpcom_tour_rate
```

**Dismiss tour by clicking on X**
```
is_gutenboarding: true
slide_number: 8
action: close-btn
_en: calypso_editor_wpcom_tour_dismiss
```

**Dismiss tour by clicking on No thanks**

```
is_gutenboarding: true
slide_number: 1
action: close-btn // <-- Should be 'no-thanks-btn'
_en: calypso_editor_wpcom_tour_dismiss
```

p-jackson reviewed (approved):
Looking great! I think this PR is g2g, I had a thought that might be more a design thing though @dubielzyk. I was thinking about how the "thumbs up" rating is persisted across all your runs through the tour, and I wondered whether maybe the button should disappear after a rating has been given.
The case I thought about was where a user might be having a tough time figuring out the editor and they go through the welcome tour, and they keep seeing this thumbs up icon, even though now they might be feeling frustrated and annoyed, and like the thumbs up icon is taunting them 😅 especially since they can't change it (the button's disabled once a rating has been given).
Maybe that's a stretch, just something I can imagine someone feeling. Even if we do change that I don't think it needs to be done in this PR.

autumnfjeld committed:
WelcomeTour: Set data store initial values to undefined

autumnfjeld committed:
WelcomeTour: bug fix: pass onDismiss function down without arguments

ramonjd reviewed (approved):
Thanks for the updates!

Can confirm that the dismiss button is working for me now

```
is_gutenboarding: false
slide_number: 1
action: no-thanks-btn
_en: calypso_editor_wpcom_tour_dismiss
```

deBhal commented:
Works for me too. The modal looks great! Thanks for the help getting there :)

I set `localStorage.setItem( 'debug', 'calypso:analytics' )` inside the `post.php` context  for my site (i.e. inside the iframe) and saw all the different events:
![Edit_Page_‹_julesaustestnuxmodal_—_WordPress_com](https://user-images.githubusercontent.com/5952255/102948405-aaf6e580-4511-11eb-8e7f-f1601d916335.jpg)

Not for this PR, but one thing I did notice was that when I hit the thumbs-down button by accident, I was not able to change it to a thumbs-up.

Edit: Also, that e2e test failure should be resolved by rebasing.

ramonjd commented:
> Not for this PR, but one thing I did notice was that when I hit the thumbs-down button by accident, I was not able to change it to a thumbs-up.

This is a valid point. What happens if I accidentally click thumbs down, but think better of it. I do that all the time!

autumnfjeld commented:
> > Not for this PR, but one thing I did notice was that when I hit the thumbs-down button by accident, I was not able to change it to a thumbs-up.
> 
> This is a valid point. What happens if I accidentally click thumbs down, but think better of it. I do that all the time!

Yes, I'll make a note of this.  I set it up deliberately like that: for instance we don't want to track both a thumbs up and a thumbs down event....that wouldn't be helpful. Thus if we wanted to allow the user to change their mind, then we would need a third button "submit" .    I think more thought needs to be put into how that rating should function and if it should also be hidden after rating.
